src/edu/washington/escience/myria/operator/agg/StreamingAggregate.java

```diff
@@ -0,0 +1,275 @@
+                break;
+              case STRING_TYPE:
+                curGroupKey.set(gKey, tb.getString(gFields[gKey], row));
+                break;
+              case DATETIME_TYPE:
+                curGroupKey.set(gKey, tb.getDateTime(gFields[gKey], row));
+                break;
+              case INT_TYPE:
+                curGroupKey.set(gKey, tb.getInt(gFields[gKey], row));
+                break;
+              case LONG_TYPE:
+                curGroupKey.set(gKey, tb.getLong(gFields[gKey], row));
+                break;
+              case FLOAT_TYPE:
+                curGroupKey.set(gKey, tb.getFloat(gFields[gKey], row));
+                break;
+              case DOUBLE_TYPE:
+                curGroupKey.set(gKey, tb.getDouble(gFields[gKey], row));
+                break;
+            }
+          }
+          reinitializeAggStates();
+        }
+        // update aggregator states with current tuple
+        for (int agg = 0; agg < aggregators.length; ++agg) {
+          aggregators[agg].addRow(tb, row, aggregatorStates[agg]);
+        }
+      }
+      tb = child.nextReady();
+    }
+
+    /*
+     * We know that child.nextReady() has returned <code>null</code>, so we have processed all tuple we can. Child is
+     * either EOS or we have to wait for more data.
+     */
+    if (child.eos()) {
+      addToResult();
+      return getResultBatch();
+    }
+    return null;
+  }
+
+  /**
+   * Re-initialize aggregator states for new group key.
+   *
+   * @throws DbException if any error
+   */
+  private void reinitializeAggStates() throws DbException {
+    aggregatorStates = null;
+    aggregatorStates = AggUtils.allocateAggStates(aggregators);
+  }
+
+  /**
+   * Add aggregate results with previous grouping key to result buffer.
+   *
+   * @throws DbException if any error
+   */
+  private void addToResult() throws DbException {
+    int fromIndex = 0;
+    for (; fromIndex < curGroupKey.numColumns(); ++fromIndex) {
+      TupleUtils.copyValue(curGroupKey, fromIndex, 0, resultBuffer, fromIndex);
+    }
+    for (int agg = 0; agg < aggregators.length; ++agg) {
+      aggregators[agg].getResult(resultBuffer, fromIndex, aggregatorStates[agg]);
+      fromIndex += aggregators[agg].getResultSchema().numColumns();
+    }
+  }
+
+  /**
+   * @return A batch's worth of result tuples from this aggregate.
+   * @throws DbException if there is an error.
+   */
+  private TupleBatch getResultBatch() throws DbException {
+    Preconditions.checkState(getChild().eos(), "cannot extract results from an aggregate until child has reached EOS");
+    if (finalBuffer == null) {
+      finalBuffer = resultBuffer.finalResult();
+      if (resultBuffer.numTuples() == 0) {
+        throw new DbException("0 tuples in result buffer");
+      }
+      resultBuffer = null;
+    }
+    if (finalBuffer.isEmpty()) {
+      return null;
+    } else {
+      return finalBuffer.get(0);
```

Comment:
So this means that you can only return a single `TupleBatch` from your `TupleBuffer` even if you have accumulated multiple `TupleBatch`es?